### ISSUE TITLE
Problems creating a TIFF using a custom color scheme and TYPE_BYTE_INDEXED BufferedImage

### ISSUE THREAD

**Issue by FelixA** [haraldk/TwelveMonkeys#306]
Hi Harald,

I've been playing further around with your library. 
One Thing I would like to do is reducing the Image size of an Image by limiting the Color Palette. 
A tif is actually being created, but the Colors and the Image itself is corrupted.

For that I'm using an IndexColorModel and create a new BufferedImage with the Type TYPE_BYTE_INDEXED as you can see here: 

```public static void reduceColorPalette(BufferedImage inputImage, File outputPath) {
		IndexColorModel cm = new IndexColorModel(3, 8,

				// RED GREEN1 GREEN2 BLUE WHITE BLACK
				// new byte[] { -100, 0, 0, 0, -1, 0 },
				// new byte[] { 0, -100, 60, 0, -1, 0 },
				// new byte[] { 0, 0, 0, -100, -1, 0 });

				// RED GREEN BLUE YELLOW 			DARKGRAY 	WHITE 	LIGHTGRAY 		BLACK

				new byte[]{ (byte) 255, 0, 0,    (byte) 255,   (byte)64,    (byte) 255,  (byte) 192, (byte) 0 }, 
				new byte[]{ 0,(byte) 255, 0,     (byte) 255,   (byte)64,   (byte) 255,  (byte) 192, (byte) 0 },
				new byte[]{ 0,  0,  (byte) 255,   0,          (byte) 64,   (byte) 255,  (byte) 192, (byte) 0 });


	
		BufferedImage img = new BufferedImage(inputImage.getWidth(), inputImage.getHeight(), BufferedImage.TYPE_BYTE_INDEXED, cm);
		Graphics2D g2 = img.createGraphics();
		g2.drawImage(inputImage, 0, 0, null);
		g2.dispose();

		// TIFUtils.writeBufferedImageToFile(outputPath, img, "LZW");

		// output
		try
		{

			ImageIO.write(img, "png", outputPath);
		}
		catch (IOException e)
		{
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

	}
```

Writing this out as a PNG works perfectly, but the custom tif writer seems to have Problems with this type of BufferedImage. The following is the code I use to write the TIFF to file:

```public static boolean writeBufferedImageToFile(File outputFileName, Object inputImage, String targetCompressionType)

	{
		if (inputImage instanceof BufferedImage || inputImage instanceof PlanarImage || inputImage instanceof RenderedImage)
		{

			buffer = new ByteArrayOutputStream();

			ImageOutputStream stream = null;

			try
			{
				stream = ImageIO.createImageOutputStream(buffer);

				TIFFImageWriterSpi imageWriterSPI = new TIFFImageWriterSpi();

				TIFFImageWriter writer = new TIFFImageWriter(imageWriterSPI);

				writer.setOutput(stream);

				out = ImageIO.createImageOutputStream(outputFileName);

				writer.setOutput(out);

				writer.prepareWriteSequence(null);

				TIFFImageWriteParam params = new TIFFImageWriteParam(Locale.GERMAN);

				// Sets Compression Mode. 2 is MODE_EXPLICIT which is needed to
				// pass the compression parameter.
				params.setCompressionMode(2);

				params.setCompressionType(targetCompressionType);

				writer.writeToSequence(new IIOImage((RenderedImage) inputImage, null, null), params);

				out.close();

				stream.close();

				buffer.close();

				writer.dispose();
			}
			catch (IOException e)
			{
				logger.error("CRITICAL: Exception while writing the image");
				try
				{
					stream.close();
				}
				catch (IOException e1)
				{
					// TODO Auto-generated catch block
					e1.printStackTrace();
				}
				e.printStackTrace();
				return false;
			}
		}
		else
		{
			logger.error(String.format("Wrong image format %s", inputImage.getClass()));
			return false;
		}
		return false;
	}
```

Would be awesome if you'd have an Idea why using this model tifs cannot be created

**Comment by haraldk** [comment]
Hi Felix,

Can you please attach a (preferably small) sample image, and the result you get? Also, what version of the library are you using? 

There seems to be some issues with your TIFF writing code above, but I don't think they are related to the problems you see.  

- Why do you have two `ImageOutputStream`s? The in-memory stream is never written to it seems.

- You should also use `ImageIO. getImageWritersByFormatName("TIFF")`, rather than creating a SPI and a writer directly. While your code works now, it will (most probably) not work with Java 9. Creating a new `TIFFImageWriteParam` is allowed, but you could also obtain one from the writer, to avoid the compile time dependency, like `writer.getDefaultWriteParam()`. 

- You are writing a sequence for a single image. While this should work fine, it makes the code harder to read and understand.

- Your method seems to always return `false`. If so, `void` is probably a better type. :-)

As I said, these issues are likely not the cause of the issue. I'll look into this in more detail when I receive some more details. 


Best regards,

-- 
Harald K

**Comment by haraldk** [comment]
Hi again Felix,

How are you doing? Do you stil have this problem? 

Thanks for reporting anyway, I found two separate issues:

When I tried to recreate a full program from your code, it seems to work for me, *if I change the compression from LZW* to pretty much any other compression that supports palette data (ie. "None", "Zlib"/"Deflate" or "PackBits"). The TIFFImageReader could then read the images back. 

However, they couldn't be opened in 3rd party tools, like Windows Photo Viewer, as there seems to be a bug (inconsistency) in the handling of IndexColorModel where bits per sample is < 8. In your case, the writer will write 8 bits/sample, as this is the sample size in the SampleModel, but the palette will contain only 24 colors, as the pixel size in the ColorModel is 3. It should to be either 8 bits/sample and 256 entries in the color table OR 3 bits/sample and 24 entries in the color table (and of course, the pixel data would have to match). The latter would of course be preferable to you, if you want the smallest possible file size. Neither the TIFFImageReader or TIFFImageWriter supports this at the moment though, but I might consider adding support for something like this in the future.

In addition have found a related issue using LZW compression in combination with IndexColorModel where bits per sample is < 8 (as above), Currently, this makes the LZWEncoder ouput bad LZW codes, which is most likely the issue you ran into. The files written in this case, will be corrupted and not readable by any software (although you might run into some lucky cases).

Again thanks! I'll try to create some test cases and fix the bugs as soon as possible.


Best regards,

-- 
Harald K

**Comment by FelixA** [comment]
Hi Harald,

thank you for the awesome reply and your suggestions regarding refactoring. I had to rush a bit to code this together, but nevertheless your feedback is very appreciated. I'm busy with other issues for this week, but next one I am able to dig into the image processing again.

So, do I understand right, that using a ColorModel with 8 bits/sample and 256 entries is supported by the current implementations of TIFFImageReader and TIFFImageWriter, 
while 3 bits/sample and 24 entries is not supported?

Well, that explains why using this algorithm for the images I'm processing didn't result in the awesome compression rates I was expecting, since I'm basically dealing with black and white images that have e.g. a colored labeling that needs to be preserved, while it is no issue if it goes from slightly red to red or s.th. like that.

One thing I wonder about is, is there is some compression algorithm like .gif that reduces the color scheme to a variable amount of colors, instead the fixed 256? Maybe it's a problem with the portability of the images then?

Also interesting that only LZW compression has issues with less than 8 bits per sample. 

Again thanks a lot also for your great explanations. Honestly I didn't expect that working with images results in such a steep learning curve!

Best regards
Felix

**Comment by haraldk** [comment]
Hi again,

Yes, there's currently two bugs related to writing TIFFs with IndexColorModel and < 8 bit/sample. One is the writing of the ColorMap entry which is broken (but can still be opened using the TIFFImageReader), and the other is related to LZW decoding. I will fix these bugs.

Also, the writer will write 8 bits/sample regardless, so you are right, the files will not be a lot smaller as you would expect (although the LZW compression should compress this data very good, had it worked...).

The PNG format is a better format for storing variable number of colors indexed color model, I guess. But if you need it to be TIFF, then using a ColorMap with less entries is your best bet, and the LZW compression is basically the same algorithm as used in GIF (not *exactly* the same, but close). Deflate compression should yield similar compression ratios as LZW, so you could use that until the LZW encoding is fixed. Using a multiple of two number of bits (2, 4 , 16 or 256 colors) will generally be more compatible with TIFF decoders. But as mentioned, using < 8 bits also requires some changes in the TIFFImageWriter.

PS: You might want to have a look at the `com.twelvemonkeys.image.IndexImage` and  `com.twelvemonkeys.image.DiffusionDither` classes for dynamic reduction of colormap entries and higher quality dithering. 

-- 
Harald K

**Comment by haraldk** [comment]
Hi Felix,

I believe I have fixed this issue now in the 762b596 commit. Are you able to test and verify the fix?


Best regards,

-- 
Harald K

**Comment by FelixA** [comment]
Hi Harald,

thank you for the fast changes. Unfortunately I'm a bit slower.
I try to verify the fixes by the end of the week.

But already, thank you very much!

Best regards,
Felix

**Comment by haraldk** [comment]
I just released 3.3.2 now, which should fix your issue. If not, please reopen or file a new issue. 

Closing for now.

-- 
Harald K
